pkg/controller/containersource/reconcile.go

```diff
@@ -111,9 +111,10 @@ func (r *reconciler) Reconcile(ctx context.Context, object runtime.Object) (runt
 	expected := resources.MakeDeployment(nil, args)
 	if !equality.Semantic.DeepEqual(deploy.Spec, expected.Spec) {
 		deploy.Spec = expected.Spec
-		if r.client.Update(ctx, deploy); err != nil {
-			return object, err
-		}
+		err := r.client.Update(ctx, deploy)
+		source.Status.MarkDeploying("DeployUpdated", "Updated deployment %s", deploy.Name)
```

Comment:
Good catch! What should happen when `err != nil` here? I'm not sure whether status should change or not.

Seems like what will happen is:
- Mark status deploying
- Ready will become False
- Immediate reconcile again
- Attempt a second update

For a transient error, this is effectively the same as only updating status on success.

In the case of persistent errors, the Deployment never actually changes so the data plane is still technically working. On the other hand, the Deployment is out of date so from the perspective of the ContainerSource, there's more left to do and that seems like Ready=Unknown.

I guess the question is about the definition of Ready=True: Is it "events are plumbed", or "reconcile is complete"?

I see the [serving documentation about conditions](https://github.com/knative/serving/blob/22aacc4a25837fb1eac408cc8925707e557b0248/docs/spec/errors.md#knative-serving-condition-status) defines Ready as the latter.

So if we're conforming to the knative/serving conditions semantics (and we should), then **the status should be updated immediately after recognizing that reconciliation is required**. That's what's happening here, so this comment can be disregarded :grin:

Reply:
This probably suggests that you need another condition to indicate whether the data plane is up to date, or you can treat this as `Ready = Unknown` until the Deployment matches the requested state.